### PR TITLE
Add constrained solving to minimize and maximize

### LINKED ISSUE #62: Add constrained solving to Julia minimize and maximize
Parent epic: #55

Planned PR branch/name: `pr-7-constrained-minimize-maximize`

## Scope
- Add `constraints=AbstractConstraint[]` keyword to matrix QUBO `minimize`, polynomial `minimize`, and `maximize`.
- Keep the unconstrained path identical when `constraints` is empty.
- In constrained mode, tensorize the objective, build projection MPOs, project the Hamiltonian, project the initial MPS, run through the DMRG backend path, and return the objective evaluated on a feasible sample.
- Integrate through the backend boundary introduced in #37 rather than editing the legacy solver flow directly.
- Add bounded feasible sampling retry with clear failure diagnostics.

## Acceptance Criteria
- [ ] Brute-force constrained QUBO tests pass.
- [ ] Returned sample is feasible for all v1 constraint types.
- [ ] Existing unconstrained QUBO/PUBO tests pass unchanged.
- [ ] Existing callback and iteration metadata behavior remains intact.

## Depends On
- #37 ([SpinGlassPEPS 3/6] Introduce solver backend interface while keeping DMRG default)
- #61 (Add P'HP Hamiltonian projection utilities)

## Sequencing Note
- This is the first CoTenN constraints issue expected to touch public solver orchestration. It should wait until #37 lands so constrained solving is implemented against the backend abstraction instead of the pre-refactor solver path.
- #56 through #61 may proceed before this issue because they are additive and mostly internal.

## Notes
Each issue is intended to be completed by one focused PR.
## Attribution
This work is inspired by the projection-MPO constraint framework in Sharma, Ritvik, Cheng Peng, Siddharth Dangwal, and Sara Achour. "CoTenN: Constrained Optimization with Tensor Networks." Proceedings of the ACM on Programming Languages 10, no. PLDI (2026): 629-655.

## Benchmark Follow-Up
- After constrained `minimize`/`maximize` is available, #66 will compare knapsack penalty-QUBO solving against the new projection-MPO constrained path.

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

- Add `constraints=` support to DMRG-backed QUBO and polynomial `minimize`, with `maximize` forwarding constraints through the existing sign-flip path.
- Lower constraints through the existing DFA/projection-MPO path, solve the projected Hamiltonian, project constrained DMRG states before callbacks/final sampling, and retry final sampling with clear diagnostics.
- Keep QUBO constraints in original variable coordinates when `preprocess=true`, add a zero-Hamiltonian projection guard, and mark the native constraint API experimental in docs.
- Add focused constrained-solver tests covering all v1 constraint types, preprocessing, constrained `maximize`, polynomial `minimize`, callbacks/stats, zero objectives, and infeasible diagnostics.

Closes #62.

## Tests

- `julia --project=test -e 'using Pkg; Pkg.instantiate()'`
- `julia --project=test -e 'push!(LOAD_PATH, pwd()); include(joinpath(pwd(), "test", "runtests.jl"))'`

## Branch Hygiene

- Base branch: `main`
- Source branch: `pr-7-constrained-minimize-maximize`
- Branch point: `origin/main` at `9f487c765b0b2a079a01e8f42ae7bbe7da7b451a`
- Stacked status: unstacked; #37 and #61 are already merged into `main`
- Prerequisites: none beyond current `main`

### CODE REVIEW

**Review by bernalde — COMMENTED** [review]
I'm the author of this PR, so GitHub does not let this account submit a formal `APPROVE` or `REQUEST_CHANGES` — posting as `COMMENT` with the same findings a maintainer review would raise.

## Summary

This is a solid, well-structured integration of the projection-MPO constraint machinery (from #56–#61) into the public `minimize`/`maximize` entry points, with good test coverage of the four v1 constraint types across QUBO, preprocessing, `maximize`, polynomial `minimize`, callbacks, zero-objective, and infeasibility diagnostics. I ran the full suite (`Aqua.jl`, doctests, `QUBODrivers.jl`, JuMP interface, and the new `Constrained solving` testset) and everything passes (50/50 new tests). However, I found one reproducible crash on an edge case that isn't covered by the new tests.

## Blocking

**Constrained solve crashes on single-variable problems with a confusing low-level error, regressing a previously-supported input size.**


`minimize_mpo` only takes the `single_variable_minimize` fast path when `isnothing(initial_state) && isnothing(solution_projections)` ([src/backends/dmrg.jl:397](src/backends/dmrg.jl#L397)). Any constrained solve always builds `initial_state`/`solution_projection` (via `constrained_projection_problem`), so a single-variable constrained problem is forced through the general DMRG sweep loop instead. `ITensorMPS.dmrg` explicitly does not support 1-site systems and throws:

```
ERROR: `dmrg` currently does not support system sizes of 1. You can diagonalize the MPO tensor directly with tools like `LinearAlgebra.eigen`, `KrylovKit.eigsolve`, etc.
```

Reproduced on current head (`c1a9e47`) with:

```julia
Q = zeros(1,1); l = [-1.0]
constraints = AbstractConstraint[SumConstraint([1],[1],1;relation=:(==))]
minimize(Q, l; constraints, iterations=3, verbosity=1, cutoff=1e-12)
```

and equivalently through the polynomial `minimize(::AbstractPolynomial; constraints, ...)` path (same `ITensorMPS.dmrg` error, one stack frame down).

This matters because single-variable QUBO/PUBO is a deliberately supported and tested case elsewhere in the repo (`test/backend.jl`, `test/qubo.jl`), and the acceptance criteria on #62 call for "Existing unconstrained QUBO/PUBO tests pass unchanged" plus "Returned sample is feasible for all v1 constraint types" — implicitly across valid problem sizes, not just n≥2. Right now a user constraining a 1-variable problem (e.g. forcing a single bit to a fixed value with `SumConstraint`/`ExactlyOneConstraint`/`NotEqualsConstraint`) gets an opaque `ITensorMPS` internal error instead of a result or a clear diagnostic.

Fix options: extend `single_variable_minimize` to accept a feasibility filter (trivial for n=1 — brute-force check `x=0`/`x=1` against `is_feasible` and pick the better feasible one, erroring if neither is feasible), or at minimum raise a clear `ArgumentError`/`ErrorException` from `minimize_mpo` when `length(sites) == 1` and constraints are present, rather than letting the cryptic `ITensorMPS.dmrg` error surface. I'd lean toward the first option since it's a small addition and keeps behavior consistent with the unconstrained path. Either way, this should ship with a regression test (n=1 QUBO and/or polynomial with a constraint that pins the single bit).

## Nonblocking

**`is_zero_mpo` is a structural, not semantic, zero-check.**


`is_zero_mpo(H) = all(iszero, H)` ([src/backends/dmrg.jl:196](src/backends/dmrg.jl#L196)) checks that every individual MPO tensor is exactly zero, which happens to hold for `tensorize`'s output when `Q`/`l` (or the polynomial) are all-zero, but wouldn't detect an operator that is zero only after bond contraction/cancellation. That's fine for the current call site (right after `tensorize`), but worth a one-line comment noting the assumption so a future caller doesn't reuse it expecting a general zero-operator test.

**`feasible_sample_retries` isn't mentioned in the `maximize` docstring.**


`maximize`'s docstring only calls out the forwarded `constraints` keyword ([src/solver.jl:172-175](src/solver.jl#L172-L175)), not `feasible_sample_retries`, even though both are equally forwarded through `kwargs`. Minor doc-completeness nit, not a functional issue.

## Linked issue (#62)

The PR functionally satisfies all four acceptance criteria on #62 (brute-force-verified constrained QUBO tests, feasibility for all v1 constraint types including the `NotEqualsConstraint` DFA-path carry-over from #89, unconstrained QUBO/PUBO tests unchanged, callback/stats behavior intact) — the checkboxes on the issue just haven't been ticked yet. It also follows through on the design notes from the issue thread: native constraints are marked experimental in the `AbstractConstraint` docstring, the lowering boundary stays `constraint -> DFA -> projection MPO -> P'HP` (no ad hoc feasible-assignment construction in the public path), and projection bond-dimension metadata (`maxlinkdim` per projection/effective Hamiltonian/initial state) is exposed via `@debug`. Modulo the n=1 gap above, `Closes #62` looks appropriate.

## Tests run

- `julia --project=test -e 'using Pkg; Pkg.instantiate()'` — clean instantiate.
- `julia --project=test -e 'push!(LOAD_PATH, pwd()); include(joinpath(pwd(), "test", "runtests.jl"))'` — full suite green, including `Constrained solving` (50/50), `Aqua.jl` (11/11), `Doctests` (1/1), `QUBODrivers.jl` (175/175).
- Manual repro scripts for the single-variable constrained QUBO and polynomial paths (see Blocking finding above) confirmed the crash; cleaned up afterward, no residue in the working tree.

## Merge readiness

I would not merge this until the blocking issue above is addressed (or explicitly scoped out with a clear error message and a note in the docs/issue that n=1 constrained solves are unsupported). Everything else is in good shape. Since I authored this PR, I can't provide the formal approval GitHub requires for merge — that needs to come from another maintainer/account once the above is resolved.

**Comment by bernalde** [comment]
Addressed the PR #94 review feedback on head `f6596aa64ed81b3b9fb843ff3fa9cadb1c0b68e4`.

Commits pushed:
- `f6596aa` `fix: handle constrained single-variable DMRG solves`

Main changes:
- `n1-constrained-dmrg-crash`: kept one-variable solves on the scalar fast path and added feasibility filtering there, so constrained n=1 QUBO and polynomial problems return feasible results instead of entering one-site `ITensorMPS.dmrg`.
- `is-zero-mpo-heuristic`: added a comment that `is_zero_mpo` is a structural check for freshly tensorized zero objectives, not a general semantic zero-operator test.
- `maximize-docstring-retries`: documented that `maximize` forwards both `constraints` and `feasible_sample_retries`.

Tests run:
- Added the n=1 constrained regression tests first; they reproduced the one-site `dmrg` crash before the fix.
- `julia --project=test -e 'push!(LOAD_PATH, pwd()); using Test, Random, LinearAlgebra; using TenSolver; import ITensors; ITensors.enable_debug_checks(); filepath(x)=joinpath(pwd(), "test", x); include(filepath("utils.jl")); include(filepath("constrained_solve.jl"))'` - passed, `68/68`.
- `julia --project=test -e 'push!(LOAD_PATH, pwd()); include(joinpath(pwd(), "test", "runtests.jl"))'` - full suite passed, including doctests; `Real Models` still reports its existing `3` broken tests.
- GitHub Actions on `f6596aa`: Documentation passed; full Julia matrix passed on Ubuntu, macOS, and Windows.

Comments intentionally not addressed:
- None.

Remaining:
- PR #94 is open, ready for review, and has clean merge state with green checks. A formal review from another maintainer/account is still required because the previous review body was posted by the PR author account.

File: src/constraints.jl [hunk 1]
```diff
@@ -12,6 +12,11 @@ Any concrete subtype is expected to implement
 - [`is_feasible`](@ref).
 - [`constraint_sites`](@ref).
 
+Native constraint types are experimental. They currently provide TenSolver's
+Julia lowering target for projection-MPO constrained solves; future JuMP/MOI
+integration may change which constraint abstraction is considered stable public
+API.
```
Comment (iagoleal) [inline src/constraints.jl]:
```suggestion
Constraint types are experimental. They currently provide TenSolver's
Julia lowering target for projection-MPO constrained solves; future JuMP/MOI
integration may change which constraint abstraction is considered stable public
API.
```

File: src/solver.jl [hunk 2]
```diff
@@ -163,6 +170,9 @@ for maximization.
     max  b'Qb + l'b + c
     s.t. b_i in {0, 1}
 
+The same `constraints` and `feasible_sample_retries` keywords accepted by
+[`minimize`](@ref) can be used to solve constrained maximization problems.
```
Comment (iagoleal) [inline src/solver.jl]:
```suggestion
All keywords accepted by [`minimize`](@ref) can also be used for maximization problems.
```

File: src/solver.jl [hunk 3]
```diff
@@ -85,6 +85,13 @@ Keyword arguments:
 - `preprocess :: Bool` - Defaults to `false`. If `true`, permute QUBO variables before constructing the MPS Hamiltonian
   so coupled variables are closer in the one-dimensional tensor order. Samples are returned in the
   caller's original variable order. This is an experimental feature and may be subject to changes.
+- `constraints :: AbstractVector{<:AbstractConstraint}` - Experimental native Julia hard constraints.
+  Defaults to `AbstractConstraint[]`. In constrained DMRG solves, TenSolver lowers each constraint to
+  a projection MPO, solves the projected Hamiltonian, and returns a feasible sampled bitstring.
+  QUBO constraints are expressed in the caller's original variable order, even when `preprocess=true`.
```
Comment (iagoleal) [inline src/solver.jl]:
```suggestion
  For polynomial objectives, constraints are expressed in its effective variable order.
```

File: src/solver.jl [hunk 4]
```diff
@@ -85,6 +85,13 @@ Keyword arguments:
...
+  Polynomial constraints are expressed in the effective variable order used by the polynomial Hamiltonian.
+- `feasible_sample_retries :: Int` - Maximum attempts to draw a feasible bitstring from a constrained
+  final state before throwing a diagnostic error. Defaults to `100`.
```
Comment (iagoleal) [inline src/solver.jl]: This is DMRG-specific and should be in the DMRG-backend documentation,

File: src/backends/dmrg.jl [hunk 5]
```diff
@@ -82,18 +101,156 @@ Solve the Polynomial Unconstrained Binary Optimization problem
...
+  return minimize_mpo(
+    H_eff,
+    cte,
+    a -> p(vs => a);
+    cutoff,
+    initial_state,
+    solution_projection=projections,
+    sample_validator=x -> is_feasible(x, constraint_vec),
+    kwargs...,
+  )
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: Same comment as for the other `minimize`.

File: src/backends/dmrg.jl [hunk 6]
```diff
@@ -82,18 +101,156 @@ Solve the Polynomial Unconstrained Binary Optimization problem
...
+
+  constraint_vec = AbstractConstraint[]
+  for (i, constraint) in pairs(constraints)
+    constraint isa AbstractConstraint ||
+      throw(ArgumentError("constraints[$i] must be an AbstractConstraint, got $(typeof(constraint))"))
+    push!(constraint_vec, constraint)
+  end
+
+  return constraint_vec
+end
```
Comment (iagoleal) [inline src/backends/dmrg.jl]:
```suggestion
function normalize_constraints(constraints)
  return collect(AbstractConstraint, constraints)
end
```

File: test/constrained_solve.jl [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    value = obj(x)
+    if value < best
+      best = value
+      best_x = x
+    end
+  end
+
+  isnothing(best_x) && throw(ArgumentError("no feasible bitstring"))
+  return best, best_x
+end
```
Comment (iagoleal) [inline test/constrained_solve.jl]: It would be better to just adapt `brute_force` to also accept constraints

File: test/constrained_solve.jl [hunk 8]
```diff
@@ -0,0 +1,214 @@
+import DynamicPolynomials
+const ConstrainedSolveDP = DynamicPolynomials
```
Comment (iagoleal) [inline test/constrained_solve.jl]:
```suggestion
import DynamicPolynomials
```

File: src/backends/dmrg.jl [hunk 9]
```diff
@@ -82,18 +101,156 @@ Solve the Polynomial Unconstrained Binary Optimization problem
...
+  )
+end
+
+function tensor_order_constraint(constraint::RelationConstraint, original_to_tensor)
+  return RelationConstraint(
+    tensor_site(constraint.left_site, original_to_tensor),
+    constraint.relation,
+    tensor_site(constraint.right_site, original_to_tensor),
+  )
+end
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: We should probably call this something like `constraint_reindex`.
The current name is confusing because these helpers are unrelated to tensors.
This logic could also be useful for other backends, so it would make sense to move the reindexing helpers to `constraints.jl`

File: src/backends/dmrg.jl [hunk 10]
```diff
@@ -82,18 +101,156 @@ Solve the Polynomial Unconstrained Binary Optimization problem
...
+    projected_hamiltonian_max_bond=ITensorMPS.maxlinkdim(H_eff),
+    projected_initial_max_bond=ITensorMPS.maxlinkdim(initial_state),
+  )
+
+  return H_eff, initial_state, projections
+end
+
+# Structural check for freshly tensorized zero objectives; this is not a
+# semantic zero-operator test after arbitrary MPO algebra.
+is_zero_mpo(H::MPO) = all(iszero, H)
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: This is fragile for Float64 coefficients.

File: src/backends/dmrg.jl [hunk 11]
```diff
@@ -225,8 +397,14 @@ function minimize_mpo( H :: MPO
                      , on_iteration :: Union{Nothing, Function} = nothing
                      , callback_every   :: Int = 1
                      , permutation :: Vector{Int} = Int[]
+                     , initial_state :: Union{Nothing, MPS} = nothing
+                     , solution_projection = nothing
+                     , sample_validator = nothing
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: Is there any case where we need a different sample_validator to is_feasible(constraints, solution)?

File: src/backends/dmrg.jl [hunk 12]
```diff
@@ -235,13 +413,25 @@ function minimize_mpo( H :: MPO
...
+      on_iteration,
+      callback_every,
+      sample_validator,
+    )
   end
 
   H     = device(H)
-  psi   = ITensorMPS.random_mps(T, sites; linkdims=inidim) |> device
+  psi   = isnothing(initial_state) ? ITensorMPS.random_mps(T, sites; linkdims=inidim) : initial_state
+  psi   = device(psi)
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: The initial state should be calculated here, not on the caller.
Currently, we do not allow warm starting.

File: src/backends/dmrg.jl [hunk 13]
```diff
@@ -82,18 +101,156 @@ Solve the Polynomial Unconstrained Binary Optimization problem
...
+
+  local last_sample
+  for _ in 1:feasible_sample_retries
+    x = sample(dist)
+    sample_validator(x) && return obj(x)
+    last_sample = x
+  end
+
+  throw(ErrorException("failed to sample a feasible constrained solution after $(feasible_sample_retries) attempts; last sample was $(last_sample)"))
+end
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: This function has many early returns, making it hard to read. It would be better to have a more structured function, perhaps using an if-else structure.

File: src/backends/dmrg.jl [hunk 14]
```diff
@@ -314,8 +515,12 @@ function minimize_mpo( H :: MPO
...
   # It makes more sense to sample a solution and calculate the true objective function applied to it.
   dist = Solution{T}(psi, energies_log, bond_dims_log, elapsed_times_log, solution_permutation)
-  x    = sample(dist)
-  optimal = obj(x)
+  optimal = sampled_objective(
+    dist,
+    obj;
+    sample_validator,
+    feasible_sample_retries,
+  )
```
Comment (iagoleal) [inline src/backends/dmrg.jl]: Theoretical question: If the Hamiltonian already contains the projections, why do we expect to encounter solutions that are not in the feasible subspace? We start in this subspace and the DMRG only sees this subspace during the optimization process.

**Review by iagoleal — CHANGES_REQUESTED** [review]

**Comment by iagoleal** [comment]
There is also a design decision to consider in here: Right now we are reporting infeasibility as an exception. This does not align with most solvers, who report feasibility as an status.
I think it would make sense to an empty Solution`and an objective of `+Inf` to report that the problem is infeasible. Moreover, the current error messages can be turned into logs.

File: src/backends/dmrg.jl (hunk 9 above)
Comment (iagoleal) [reply src/backends/dmrg.jl]: Also, it seems very wasteful to recreate all constraints. Specially since they don't really care for site order. I wonder if there is a way to make the projection_mpo construction permutation-aware.

**Comment by bernalde** [comment]
## Addressed review feedback (head `71fcf83`)

Both threads that were still unresolved were from @iagoleal. The concrete requests from that review (rename, move to `constraints.jl`, `brute_force` accepting constraints, the `Float64` sum fragility, etc.) were already handled by the refactoring commits (`4a99a0f`, `a2f5a28`, `ae833e6`); this update covers the two threads that remained open.

### Commit pushed
- `71fcf83` — fix: select feasible basis state in constrained n=1 DMRG solves

### Main change
Investigating the thread-2 question (why re-project when the Hamiltonian already contains the projections?) surfaced a real correctness bug in the n=1 constrained fast path:

- `groundstate` (n=1) picked the global lowest-energy basis state of the projected Hamiltonian `P'HP`. Because `P'HP` assigns **zero energy to the infeasible subspace** (the kernel of the projections), a feasible assignment with a *positive* objective was rejected in favour of the zero-energy infeasible state, and the following re-projection threw a misleading `constraints may be infeasible` error even though the problem was feasible.
- Reproduced on `ae833e6`: `minimize(fill(3.0,1,1); constraints=[SumConstraint([1],[1],1;relation=:(==))])` and the polynomial analogue (`p = 2z`, `ExactlyOneConstraint([1],1)`) both threw. The negative-objective cases already covered by the tests passed, which is why it slipped through.
- Fix: the n=1 path now restricts its search to feasible basis states (raising a clear `ArgumentError` only when the feasible subspace is genuinely empty). A comment now documents why the per-iteration re-projection is required.

### Tests
- Added positive-objective single-variable regression tests (QUBO and polynomial) to `test/constrained_solve.jl`. Confirmed they throw on the pre-fix code and pass after the fix.
- Full suite green locally (`julia --project=test -e 'push!(LOAD_PATH, pwd()); include(joinpath(pwd(),"test","runtests.jl"))'`): 0 failures/errors, including `Constrained solving`, `Aqua.jl` (11/11), `Doctests` (1/1), `QUBODrivers.jl` (171/171). CI is running on the new head.

### Not changed (with reason)
- The `constraint_reindex` thread: the rename and the move into `src/constraints.jl` are already done. The follow-up musing about making `projection_mpo` construction permutation-aware (to avoid recreating constraints) is left as-is — the reindex only fires under a non-identity `preprocess=true` permutation, rebuilds a small constraint list once per solve, and keeps permutation handling in one place. Threading the permutation through the DFA/projection layer would spread that concern for negligible savings versus the solve. Flagged inline as a possible future optimization.

### Approval gate
@iagoleal's review remains formally `CHANGES_REQUESTED`; that stays until they re-review or dismiss it. No Blocking comments were declined.

File: src/backends/dmrg.jl (hunk 9 above)
Comment (bernalde) [reply src/backends/dmrg.jl]: The rename to `constraint_reindex` and the move into `src/constraints.jl` are done (in the later refactoring commits).

On recreating the constraints: I left the reindex as-is. It only runs under a non-identity permutation (`preprocess=true`) — `constraint_reindex` short-circuits and returns the input unchanged otherwise — and when it does run it rebuilds a small constraint list once per solve, which is negligible next to the DMRG solve. Making `projection_mpo`/`constraint_to_dfa` permutation-aware would push permutation handling down into the DFA layer for marginal savings, whereas the current approach keeps that concern in one spot. Happy to file it as a follow-up optimization if you'd prefer.

File: src/backends/dmrg.jl (hunk 14 above)
Comment (bernalde) [reply src/backends/dmrg.jl]: Good question — and chasing it down found a real bug.

For the general (n≥2) path you're right: starting feasible and applying `P'HP`, the local eigensolver stays inside the feasible subspace in exact arithmetic. The re-projection is there because the injected `noise` term and SVD truncation can leak amplitude into the infeasible subspace, and that subspace is the kernel of the projections where `P'HP` has *zero* energy — so nothing pushes the leaked amplitude back out on its own. Re-projecting each iteration guarantees the sampled bitstring is feasible (it is also the only feasibility mechanism in the zero-objective case, where `H` is left unprojected). I added a comment at that site spelling this out.

The n=1 fast path did not respect this: it enumerated the basis states and picked the global lowest-energy one on `P'HP`. Because the infeasible state has energy 0 there, a feasible assignment with a positive objective was rejected and the re-projection then threw a misleading `constraints may be infeasible` error. Fixed in `71fcf83` by restricting the n=1 search to feasible basis states, with regression tests for positive-objective single-variable QUBO/polynomial solves.

**Review by iagoleal — APPROVED** [review]

**Comment by bernalde** [comment]
> Right now we are reporting infeasibility as an exception. This does not align with most solvers, who report feasibility as a status. I think it would make sense to [return] an empty `Solution` and an objective of `+Inf` … the current error messages can be turned into logs.

Agreed. We surveyed how the Julia optimization ecosystem handles this, and status-based reporting is the convention essentially everywhere:

- **JuMP/MOI** (HiGHS, Gurobi, GLPK, SCIP, Ipopt, …): `optimize!` never throws on an infeasible model. The solver reports `termination_status == MOI.INFEASIBLE` with `result_count == 0`; the *only* thing that throws is querying a result that doesn't exist (`value`, `objective_value`). Solve reports, query throws.
- **QUBODrivers/QUBOTools samplers** (our own JuMP surface): status travels in the `SampleSet` metadata — our `tensolver_status` already emits `MOI.LOCALLY_SOLVED` / `ITERATION_LIMIT` / `TIME_LIMIT`; `MOI.INFEASIBLE` is a natural fourth branch.
- **Native-API solvers** (Optim.jl, Krylov.jl, NLsolve): return result objects with converged/solved flags rather than throwing.

The `+Inf` objective is also the mathematically honest value: the infimum over an empty feasible set is `+∞`. So let's do exactly what you propose. Concrete design:

**Native layer**
- Proven infeasibility (all-zero projection MPO at `src/backends/dmrg.jl:126-135`, or no feasible candidates in the n=1 path at `:373-378`) stops throwing: `minimize` returns `(T(Inf), infeasible Solution)` and logs `@warn` with the current message text. `maximize` gets `-Inf` for free since it is implemented as `-minimize(-p)`.
- Representation: make `Solution.tensor::Union{ITensorMPS.MPS, Nothing}` with empty stats vectors, plus an exported `is_infeasible(::Solution)` predicate. `sample` on an infeasible solution throws a clear error — mirroring JuMP's "solve reports, query throws" split.
- Scope guard: only *feasibility outcomes* become status. Genuine input errors (malformed constraints, non-integer weights, bad kwargs) stay exceptions.

**MOI/QUBODrivers layer** (ready for #65)
- `tensolver_status` gains the `MOI.INFEASIBLE, "infeasible"` branch and the sampler returns an empty `SampleSet` (`result_count == 0`), so JuMP users see exactly what every other MOI solver gives them.

**Tests/docs**
- The `@test_throws ArgumentError` infeasibility assertions in `test/constrained_solve.jl` flip to `E == Inf` / `is_infeasible` checks (this also settles the open question flagged on #63); docstrings for `minimize`/`maximize`/`Solution` document the status contract.

If this sounds right, we'll open a small tracked issue and implement it as the next constraints-tail PR — it gates the remaining #63 assertions and the #65 MOI design.

*Survey and design based on repo state as of 2026-07-13 (`main` @ 9fcc479).*
